[wip] util: Improve evaluation of includeconf lines in network sections of the config file

AkioNak opened the pull request:
This PR intends to make it easy to understand how the configuration files are interpreted.

1. Evaluate "includeconf" at the position described in the config file (like #include directive in C/C++), rather than at the end.

2. If once a section is specified with square brackets like [main], the only way to configure another network is to use square brackets again, e.g. [test].

3. In the "included" config file, the section is taken over from the "including" file. If no sections are specified yet, using square brackets or section prefix is still possible in the "included" file.

4. Changing the section in the "included" config file does not affects the including file. The section of the including file is kept.

5. If the "included" config file is included after the section has been specified by using square brackets or if they are included by using section prefix, any square brackets can not be used in that file.

ex) With Following 3 files, the properties are read as: 
```
    port=8444, main.port=8445, test.port=8442, regtest.port: undefined
```

 bitcoin.conf
 ---
 ```
test.includeconf=inc1.conf
includeconf=inc2.conf
port=8441   # ignored (2nd appearance of w/o section)
```

 inc1.conf
 ---
```
 port=8442   # test
 [regtest]
 port=8443   # ignored
```

 inc2.conf
 ---
```
 port=8444   # w/o section
 [main]
 port=8445   # main
 [regtest]
```

AkioNak committed:
util: Improve evaluation of includeconf lines

Improve property evaluation way in bitcoin.conf

This PR intends to make it easy to understand how the configuration
files are interpreted.

1. Evaluate "includeconf" at the position described in the config file
(like #include directive in C/C++), rather than at the end.

2. If once a section is specified with square brackets like [main],
the only way to configure another network is to use square brackets
again, e.g. [test].

3. In the "included" config file, the section is taken over from
the "including" file. If no sections are specified yet, using square
brackets or section prefix is still possible in the "included" file.

4. Changing the section in the "included" config file does not affects
the including file. The section of the including file is kept.

5. If the "included" config file is included after the section has been
specified by using square brackets or if they are included by using
section prefix, any square brackets can not be used in that file.

ex) With Following 3 files, the properties are read as:
port=8444, main.port=8445, test.port=8442, regtest.port: undefined

bitcoin.conf
---
test.includeconf=inc1.conf
includeconf=inc2.conf
port=8441   # ignored (2nd appearance of w/o section)

inc1.conf
---
port=8442   # test
[regtest]
port=8443   # ignored

inc2.conf
---
port=8444   # w/o section
[main]
port=8445   # main
[regtest]

promag commented:
This can break existing configurations correct?

AkioNak commented:
@promag Thanks for your comment.
Yes. Sometimes this may be different from previous results when interpreting existing configurations.
Especially the possibility of interpretation unmatch increases if using "includeconf" and sections are specified with square brackets in the "included" configuration file.
Is it necessary for this change to add a flag to specify how to read the configuration file?

practicalswift commented on src/util/system.cpp:
```diff
@@ -856,22 +857,50 @@ static bool GetConfigOptions(std::istream& stream, std::string& error, std::vect
+                if (key == "includeconf") {
+                    if (depth < 0) {
+                        continue;
+                    } else if (depth < MAX_INCLUDECONF_DEPTH) {
```
Nit: The `else` here is redundant.

AkioNak replied:
Done.

promag commented:
You should not break existing configurations, especially if there is no strong reason to do it.

dougstrong77 commented:
I need advise on some things that accurd

On Dec 4, 2018 6:40 AM, "João Barbosa" <notifications@github.com> wrote:

This can break existing configurations correct?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/14866#issuecomment-444122125>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/Abg1Yg13FWjyb49Ao07U5P5x-DvghQxAks5u1ollgaJpZM4ZATnk>
.

laanwj commented:
> @promag The includeconf and sections features are both relatively recent

I think that's a good point, does this *only* affect use of `-includeconf`?

kallewoof commented:
@laanwj I believe so, yes, except for maybe the minor detail (item 4 in the OP list). @AkioNak?

AkioNak commented:
Rebased but travis failed. Hmm,,, I will check what's going on.

AkioNak commented:
The direct cause of failing travis was potential deadlock detection. 
The investigation will continue ...

potential_deadlock_detected() reports following message:
```
POTENTIAL DEADLOCK DETECTED
Previous lock order was:
 (1) cs_args  util/system.cpp:921
 (2) csPathCached  util/system.cpp:790
Current lock order is:
 (2) csPathCached  util/system.cpp:790
 (1) cs_args  util/system.cpp:531
```

AkioNak commented:
Addressed to the potential deadlock by narrowing the range of LOCK(cs_args).

kallewoof commented:
utACK cca8a9196f491aa8e99d18846578307cd7004c45

AkioNak commented:
squashed and rebased.

Sjors commented:
Concept ACK, except I think the following example (from the description) should just fail:

```ini
[main]
includeconf=inc1.conf
test.includeconf=inc2.conf
```

Once a network is specified with square brackets like `[main]`, the only way to configure another network should be to use square brackets again, e.g. `[testsnet]`. That rule should also apply to anything included by `includeconf`.

It's important to simplify this behavior sufficiently so we stand a chance of one day merging #11082 :-)

AkioNak commented:
@Sjors Thank you for your surggestions.

> Once a network is specified with square brackets like `[main]`, the only way to configure another network should be to use square brackets again, e.g. `[testsnet]`. That rule should also apply to anything included by `includeconf`.

Good idea. I agree.

> Or, slightly less strict, that square brackets may not exist in `includeconf` files if they are included after a square bracket.

I like this. And I think square brackets may not exist in the files that are included by using "section.includeconf" style, too.

Sjors commented:
Ok, looking forward to the new version. I think it would be very useful to add a lot more tests to [feature_config_args.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_config_args.py), given that you were able to make these changes without breaking any existing test.

AkioNak commented:
rebased. (no functional change yet)

AkioNak commented:
> Once a network is specified with square brackets like [main], the only way to configure another network should be to use square brackets again, e.g. [testsnet]. That rule should also apply to anything included by includeconf.

Added rules for how to use square brackets & preriods.
Also added some tests.

@Sjors Could you review this?

AkioNak commented:
Squashed.
(Also updated commit log)

AkioNak commented:
Rebased.

AkioNak commented:
Rebased.

jtimon commented:
I don't like that in the example port=8441 in bitcoin.conf is applied only to regtest just because inc2.conf ended with [regtest].
I feel like I could get crazy for a few hours just by forgetting inc2.conf ends with [regtest] if, let's say, it previously ended with [main] but I moved chain sections around in inc.conf or something.

Looking at the example, I just can't avoid wondering why the user can't just add [regtest] to bitcoin.conf like this:

```
test.includeconf=inc1.conf
includeconf=inc2.conf
[regtest]
port=8441   # regtest
```

or just put the regtest port in inc2.conf like this.

```
port=8444   # w/o section
[main]
port=8445   # main
[regtest]
port=8441   # regtest
```

I've personally never used includeconf myself, so perhaps I just don't understand how people are using this feature.
 
Perhaps this can be done in another PR, but I feel like after includeconf, you should go back to whatever section you were in bitcoin.conf, not remain in the last one set by the included file.

AkioNak commented:
@jtimon Thank you  for the comment!

> after includeconf, you should go back to whatever section you were in bitcoin.conf, not remain in the last one set by the included file.

It makes sense. That behavior is natural so I will address.

AkioNak commented:
@jtimon I have addressed.
(Changing section in the included file never affect to the including file.)

AkioNak commented:
Rebased.

Sjors commented:
Concept re-ACK. The latest set of 5 criteria in the PR descriptor look very reasonable.

Needs rebase after @ryanofsky's settings merge refactor in #15934 (which also added lots of useful tests). I think the increased sanity warrants (if needed) breaking backwards compatibility.

ryanofsky commented:
Concept ACK, I think. But I am having a trouble understanding the description, because it seems to be describing a lot of current behavior (like points 2 and 4) that the PR doesn't actually affect, and not be clear about what is changing.

Would the following be an accurate description of what the PR does?

> Included configurations from `includeconf` lines in `[main]` `[test]` and `[regtest]` sections of the config file are now evaluated in context of that section and not treated like top-level includes. Config files included within a section are also now disallowed from changing settings in other sections.

If that is accurate, it would also be helpful to add it as an entry to: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md

The documentation https://github.com/bitcoin/bitcoin/blob/master/doc/bitcoin-conf.md should also be updated. I think most of the text in the current PR description should be removed and incorporated into that file instead.

I would also change the title of the PR to better describe what it is doing. Instead of "Improve property evaluation way in bitcoin.conf" maybe "util: Improve evaluation of includeconf lines in network sections of the config file"

AkioNak commented:
@Sjors @ryanofsky Thank you for your comments/suggestions.
I will rebase and also address the suggestions above.

adamjonas commented:
@AkioNak pinging for a rebase and addressing the comments by ryanofsky and Sjors.

AkioNak commented:
@adamjonas I will do.

AkioNak commented:
Rebased.
I am still working on addressing the comments by ryanofsky and Sjors.

adamjonas commented:
Hi @AkioNak - friendly monthly ping.

AkioNak commented:
@adamjonas Thanks. I will do again.

AkioNak commented:
I will solve the conflicts.

fanquake commented:
Going to close with "Up for Grabs".
